src/Analysis/Ast/Impl/Types/PythonFunctionOverload.cs

```diff
@@ -166,23 +167,38 @@ internal void SetReturnValueProvider(ReturnValueProvider provider)
             }
 
             // If function returns generic, determine actual type based on the passed in specific type (self).
+            // If there is no self and no declaring type, the function is standalone.
+            if (self == null && StaticReturnValue.IsGeneric() && Parameters.Any(p => p.IsGeneric)) {
+                return null; // Evaluate standalone generic with arguments instead.
+            }
             if (!(self is IPythonClassType selfClassType)) {
                 return StaticReturnValue;
             }
 
             var returnType = StaticReturnValue.GetPythonType();
             switch (returnType) {
-                case PythonClassType cls when cls.IsGeneric(): {
-                        // -> A[_T1, _T2, ...]
-                        // Match arguments 
-                        var typeArgs = selfClassType.GenericParameters.Keys
+                case PythonClassType cls when cls.IsGeneric():
+                    // -> A[_T1, _T2, ...]
+                    // Match arguments
+                    IReadOnlyList<IPythonType> typeArgs = null;
+                    var classGenericParameters = selfClassType.GenericParameters.Keys.ToArray();
+                    if (classGenericParameters.Length > 0) {
+                        // Declaring class is specific and provides definitions of generic parameters
+                        typeArgs = classGenericParameters
```

Comment:
What if there is a mix of class generic parameters and method generic parameters?

Reply:
Its OK. Specific class stores dictionary of the original generic parameter names (such as K, V in Dict) and their assigned specific types (int, str) if Dict[K, V] was created via 

```python
from typing import Dict

d1 = {1:'a', 2:'b'}
d2 = Dict(d1))
```
So here we are trying to match names of the generic function parameters to those of the class so we can figure out which specific type to use. Generally speaking method generic names must either be different or they match the class. Per Python `T` is basically a variable with the assigned type declared in the class scope. So function that is generic of T will be using class T.

Reply:
I've been thinking of a case when function has a different generic parameter. Something like this:
```python
from typing import TypeVar, Generic, Callable

T = TypeVar('T')
V = TypeVar('V')

class C(Generic[T]):
    def __init__(self, value: T) -> None:
        self.value = value

    def f(self, input: V, callback: Callable[[T, V], str]) -> str: 
        return callback(self.value, input)

def g(s1: int, s2: str) -> str:
    return str(s1) + s2

print(C(1).f("b", g))
```

Reply:
In this particular case it doesn't matter since annotation says `str` and hence

![image](https://user-images.githubusercontent.com/12820357/53542650-45cc3780-3ad4-11e9-85d5-e796474038c8.png)

Now, if we are to eval this (we don't support `Callable` just yet though, that's #535)... it is still a string since eval with call `callback` and either use annotation or eval with arguments and yield a string. 

There are probably some convoluted examples, but I'd wait for users to give them to us :-)

Reply:
Sorry, I meant `def f(self, input: V, callback: Callable[[T, V], V]) -> V` in the example above. But since we have separate task for it, that's not for this PR.